crates/libs/windows/src/core/hstring.rs

```diff
@@ -128,7 +107,23 @@ impl Clone for HSTRING {
 
 impl Drop for HSTRING {
     fn drop(&mut self) {
-        self.clear();
+        if self.is_empty() {
+            return;
+        }
+
+        unsafe {
+            let header = std::mem::replace(&mut self.0, core::ptr::null_mut());
+            // This flag indicates a "fast pass" string created by some languages where the
+            // header is allocated on the stack. "Fast pass" strings can only ever be used
+            // as non-owned input params and as such `drop` will never be called. This check
+            // just ensures the Rust translation never accidentally modeled "fast pass" input
+            // params as owned `HSTRING` types which would be a bug in the translation.
```

Comment:
Feel free to find your own words for this, or ignore altogether, but I was very confused by the use of "owned" and "params" in a context that isn't strictly dealing with either. (Perhaps also a result of my Rust naivete.)

```suggestion
            // This check ensures we do not attempt to free a "fast pass" HSTRING.
            // "Fast pass" HSTRINGs do not own the memory they describe and as such `drop`
            // should never be called.
```

Reply:
Oh, looking at the original, I see I unwittingly brought us full circle. 🤭

Reply:
😊 Unfortunately, we're dealing with APIs and callers that exist outside of Rust's notions of ownership and borrowing, but that will interact with it which makes this confusing. What's more is that we still haven't fully figured out how we map the outside world on to Rust semantics. What we currently have works, but it's inconsistent and sometimes hard to reason about.

In this particular case, the `HSTRING` type defined here in Rust really represents an `HSTRING` that owns its memory and should have its ref count decremented (and memory possibly freed) when it's no longer needed. However, `HSTRING`s in other languages don't always represent handles to memory that is owned - one can also pass around `HSTRING`s that are just meant as temporary references (and thus the ref count should be left alone). One example of this are these "fast pass" strings which (because their headers are stack allocated) can only ever be used as "in" arguments to functions. When these types of `HSTRING`s come into a function, the function must take care to not decrement the ref count, and if the function wants to store the string somewhere that lives past the function call, it must be sure to increment the ref count. 

In our Rust translation we're currently modeling this as `ManuallyDrop<HSTRING>` because essentially these types of `HSTRING`s act just like the `HSTRING`s that own their memory with the exception that when you're done with them you shouldn't decrement the ref count (which in Rust happens in the `Drop` impl. `ManuallyDrop` accomplishes this by essentially turning off the `Drop` impl for the wrapped type. This is not particularly elegant or obvious, but it works. 

This comment is meant to say that if `Drop::drop` gets called on a "fast pass" string, the windows crate has a bug in how code got generated.

It's clear we still have some work to do to make some of this clearer in the code. We have a working solution, but we have some ways to go to make that solution obvious and understandable.